### ISSUE TITLE
Infinite loading game bar when leaving a remote game

### ISSUE THREAD

**Issue by RoiEXLab** [triplea-game/triplea#3468]
Noticed this while testing the other chat issues.
Steps to reproduce:
Join a bot
Start the game
Press X to close the window or press File > Leave Game
Get a loading dialog that won't close what so ever.

If you press File > Exit Game everything works fine.

**Comment by prastle** [comment]
Yup I just closed it with task manager

**Comment by RoiEXLab** [comment]
@prastle Hmm I can no longer reproduce this exact behaviour.
It still shows a loading bar, but when clicking cancel on the ClientOptions window thingy it closes the loading screen as well.
Perhaps a side effect of #3467 ?

**Comment by Cernelius** [comment]
I cannot reproduce this with latest, either. Was guessing I was just being lucky.

**Comment by RoiEXLab** [comment]
Thanks for the second confirmation. @Cernelius
Closing this then.

@ron-murhammer Now we should have a stable version candidate

**Comment by Cernelius** [comment]
Just to clarify. I never had this issue (probably because I skipped that prerelease). So, maybe, you want to wait for someone that used to have it and not anymore, as a true confirmation that is gone.

**Comment by prastle** [comment]
I have not been playing much online so I can't confirm. @Cernelius @ cer join and we test whatever pre release you think it is fixed in.

**Comment by prastle** [comment]
It is still occurring with the latest as well as this error @RoiEXLab 
When attempting to use toolbox to view a bot host.

TripleA engine version 1.9.0.0.10355
Loading map: , from: C:\Users\prast\triplea\downloadedMaps\
Loading resources from the following paths: [C:\Users\prast\triplea\downloadedMaps\, G:\10355\TripleA\assets]
Error: Exception on remote
java.lang.RuntimeException: Exception on remote
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:54)
	at com.sun.proxy.$Proxy2.getHostConnections(Unknown Source)
	at games.strategy.engine.lobby.client.ui.LobbyGamePanel.getHostInfo(LobbyGamePanel.java:275)
	at games.strategy.engine.lobby.client.ui.LobbyGamePanel.lambda$getGeneralAdminGamesListContextActions$6(LobbyGamePanel.java:196)
	at games.strategy.ui.SwingAction$1.actionPerformed(SwingAction.java:65)
	at javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
	at javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.setPressed(Unknown Source)
	at javax.swing.AbstractButton.doClick(Unknown Source)
	at javax.swing.plaf.basic.BasicMenuItemUI.doClick(Unknown Source)
	at javax.swing.plaf.basic.BasicMenuItemUI$Handler.mouseReleased(Unknown Source)
	at java.awt.AWTEventMulticaster.mouseReleased(Unknown Source)
	at java.awt.Component.processMouseEvent(Unknown Source)
	at javax.swing.JComponent.processMouseEvent(Unknown Source)
	at java.awt.Component.processEvent(Unknown Source)
	at java.awt.Container.processEvent(Unknown Source)
	at java.awt.Component.dispatchEventImpl(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.processMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.dispatchEvent(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Window.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)
Caused by: games.strategy.engine.message.RemoteNotFoundException: Not found:games.strategy.engine.lobby.server.RemoteHostUtils:Bot201_CA_USA_lobby_watcher (1)_lobby_watcher port:59574 ip:173.255.243.147
	at games.strategy.engine.message.UnifiedMessengerHub.messageReceived(UnifiedMessengerHub.java:89)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.send(UnifiedMessenger.java:220)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWaitRemote(UnifiedMessenger.java:123)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWait(UnifiedMessenger.java:100)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:52)
	at com.sun.proxy.$Proxy4.getConnections(Unknown Source)
	at games.strategy.engine.lobby.server.ModeratorController.getHostConnections(ModeratorController.java:352)
	at sun.reflect.GeneratedMethodAccessor35.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeSingle(EndPoint.java:157)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeMultiple(EndPoint.java:140)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeLocal(EndPoint.java:126)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.lambda$messageReceived$1(UnifiedMessenger.java:275)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: remote not found
	at java.lang.Thread.getStackTrace(Thread.java:1559)
	at games.strategy.engine.message.MessengerException.fillInInvokerStackTrace(MessengerException.java:19)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:56)
	... 12 more
Error: Exception on remote
java.lang.RuntimeException: Exception on remote
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:54)
	at com.sun.proxy.$Proxy2.getHostConnections(Unknown Source)
	at games.strategy.engine.lobby.client.ui.LobbyGamePanel.getHostInfo(LobbyGamePanel.java:275)
	at games.strategy.engine.lobby.client.ui.LobbyGamePanel.lambda$getGeneralAdminGamesListContextActions$6(LobbyGamePanel.java:196)
	at games.strategy.ui.SwingAction$1.actionPerformed(SwingAction.java:65)
	at javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
	at javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.setPressed(Unknown Source)
	at javax.swing.AbstractButton.doClick(Unknown Source)
	at javax.swing.plaf.basic.BasicMenuItemUI.doClick(Unknown Source)
	at javax.swing.plaf.basic.BasicMenuItemUI$Handler.mouseReleased(Unknown Source)
	at java.awt.AWTEventMulticaster.mouseReleased(Unknown Source)
	at java.awt.Component.processMouseEvent(Unknown Source)
	at javax.swing.JComponent.processMouseEvent(Unknown Source)
	at java.awt.Component.processEvent(Unknown Source)
	at java.awt.Container.processEvent(Unknown Source)
	at java.awt.Component.dispatchEventImpl(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.processMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.dispatchEvent(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Window.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)
Caused by: games.strategy.engine.message.RemoteNotFoundException: Not found:games.strategy.engine.lobby.server.RemoteHostUtils:Bot201_CA_USA_lobby_watcher (1)_lobby_watcher port:59574 ip:173.255.243.147
	at games.strategy.engine.message.UnifiedMessengerHub.messageReceived(UnifiedMessengerHub.java:89)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.send(UnifiedMessenger.java:220)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWaitRemote(UnifiedMessenger.java:123)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWait(UnifiedMessenger.java:100)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:52)
	at com.sun.proxy.$Proxy4.getConnections(Unknown Source)
	at games.strategy.engine.lobby.server.ModeratorController.getHostConnections(ModeratorController.java:352)
	at sun.reflect.GeneratedMethodAccessor35.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeSingle(EndPoint.java:157)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeMultiple(EndPoint.java:140)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeLocal(EndPoint.java:126)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.lambda$messageReceived$1(UnifiedMessenger.java:275)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: remote not found
	at java.lang.Thread.getStackTrace(Thread.java:1559)
	at games.strategy.engine.message.MessengerException.fillInInvokerStackTrace(MessengerException.java:19)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:56)
	... 12 more

**Comment by RoiEXLab** [comment]
@prastle I assume you're talking about 2 different issues, right? 
In this case I'd ask you to create a separate issue.

**Comment by prastle** [comment]
<img width="957" alt="2018-06-23" src="https://user-images.githubusercontent.com/17258293/41812407-3e41dedc-76f0-11e8-9bd6-394c0205ff80.png">
@RoiEXLab   Nope they seem similar to me since I also get the Null Java error if attempting to join.

**Comment by RoiEXLab** [comment]
You were talking about the toolbox, which made me curious.
Weirdly enough I don't get any errors when joining whatsoever.
I can just play normally without any problems

**Comment by prastle** [comment]
it is only some bots
but can confirm toolbox broken if java null pointer error

**Comment by Cernelius** [comment]
Tried with all the bots having 0 players inside (for total 9), with 1.9.0.0.10402; didn't have this (or any) issues.
So I guess something that doesn't happen to me, but only to some.

**Comment by prastle** [comment]
@Cernelius It only occurs if the bot is occupied and you try to join it.  Also only occurs with some of them.

**Comment by Cernelius** [comment]
Then, I suggest modifying the first post or opening a new issue. What I read at the moment is:

> Noticed this while testing the other chat issues.
> Steps to reproduce:
> Join a bot
> Start the game
> Press X to close the window or press File > Leave Game
> Get a loading dialog that won't close what so ever.
> If you press File > Exit Game everything works fine.

**Comment by RoiEXLab** [comment]
@Cernelius exactly That's why I was confused as well.

@prastle Please open another issue and put your posts fron here into it

**Comment by prastle** [comment]
The "Infinite loading game bar when leaving a remote game." still exists in 10405.
<img width="546" alt="2018-06-24 infinite loading game" src="https://user-images.githubusercontent.com/17258293/41819665-5084fca8-7792-11e8-9741-55fe483483af.png">

**Comment by Cernelius** [comment]
Probably you should also remove the "close pending confirmation" label then...

Anyway, if it happens to me too, I will let you know.

**Comment by Cernelius** [comment]
I just left Bot502_LONDON_UK (was an observer of a game already in progress) and got stuck upon an endless loading bar that is telling me to please wait forever (1.9.0.0.10402).

So, unless any recent changes, you can safely remove the confirmation label now.

**Comment by Cernelius** [comment]
I downloaded 1.9.0.0.10440 right now and joined and exited the sam bot with the same game in progress: same endless loading bar.
So, if a developer is around, here is your chance to try to reproduce it.
I confirm you can remove the confirmation label.

**Comment by RoiEXLab** [comment]
@prastle @Cernelius still couldn't reproduce the issue, however I believe this issue should be fixed in 1.9.0.0.10446
Please confirm if this issue is resolved for you as well.

**Comment by Cernelius** [comment]
Issue confirmed, with 1.9.0.0.10467, for leaving, as guest, a game in progress in Bot203_CA_USA.

Keep try. Good luck.

**Comment by Cernelius** [comment]
Also, why not labelling this both as "Bug" and as "Release Blocker". I guess you assume a "Release Blocker" must be a bug, but wouldn't that make more sense, and allow to actually get all bugs via the label?

**Comment by RoiEXLab** [comment]
@Cernelius Release Blocker is a type of bug that prevents making a new release stable.
If it's just a weird visual thing that you can just ignore and continue playing then this would be a bug but not a release blocker

**Comment by Cernelius** [comment]
But I want to point out again that this (at the first post of this issue) doesn't actually happen to me:

> Join a bot
> Start the game
> Press X to close the window or press File > Leave Game
> Get a loading dialog that won't close what so ever.

Just leaving a game in progress as a guest, so far. Not sure I am being off topic, as I pointed out.

**Comment by prastle** [comment]
@RoiEXLab I sent the latest thread dump.  Let me know if I did it right.

**Comment by RoiEXLab** [comment]
@prastle Perfectly 👍
This one looks a lot better than the other one.

**Comment by RoiEXLab** [comment]
@prastle Is the second Thread Dump from the right process?
Needs to be the PID of GameRunner, not TripleA.exe

**Comment by prastle** [comment]
Re did and sent you the latest in case I did it wrong first time.  It is the game runner 100%

**Comment by RoiEXLab** [comment]
@prastle @Cernelius Could you try out 1.9.0.0.10480 and see if it solved the issue?

**Comment by Cernelius** [comment]
Ok so why closing if we still have to test/confirm? I was oriented not to, since now it is closed, but I'll try now, since you specifically asking.
Shouldn't be open with the confirmation label?

**Comment by RoiEXLab** [comment]
Closing is automatic when the other PR gets merged, so...

**Comment by Cernelius** [comment]
Ah ok, good to know. So far I never had the abit to even look at closed stuff, except very rarely.

**Comment by prastle** [comment]
Yup works perfectly ther than the confirm window which should not be there could job sir
@RoiEXLab

**Comment by Cernelius** [comment]
I joined and left all games currently in progress in the lobby, without having any issues upon leaving.

Anyways, in the time that it takes to join, I get the regular initial TripleA screen, and I can even click to "Start Local Game" or whatever in it, that I'm thinking I should not get the chance to, since I'm actually waiting to join a game.

For example if, on that screen, while waiting, I, for whatever reasons, decide to play a local game, this is what happens:
![20180701_02](https://user-images.githubusercontent.com/19562310/42137372-7e91e618-7d6b-11e8-9042-eca06b8eb794.png)

And I get this error log:
```
TripleA engine version 1.9.0.0.10480
Loading map: 270bc_variants, from: C:\Users\001\Documents\GitHub\270bc_variants\map
Loading resources from the following paths: [C:\Users\001\Documents\GitHub\270bc_variants\map, C:\Program Files\TripleA\assets]
Loading map: 270bc_variants, from: C:\Users\001\Documents\GitHub\270bc_variants\map
Loading resources from the following paths: [C:\Users\001\Documents\GitHub\270bc_variants\map, C:\Program Files\TripleA\assets]
Loading map: world_war_ii_global, from: C:\Users\001\Documents\GitHub\world_war_ii_global-master.zip
Loading resources from the following paths: [C:\Users\001\Documents\GitHub\world_war_ii_global-master.zip, C:\Program Files\TripleA\assets]
Loading map: world_war_ii_global, from: C:\Users\001\Documents\GitHub\world_war_ii_global-master.zip
Loading resources from the following paths: [C:\Users\001\Documents\GitHub\world_war_ii_global-master.zip, C:\Program Files\TripleA\assets]
Could not find: null/doc/images/Teams.png
Could not find: null/doc/images/01archer01.png
Could not find: null/doc/images/01archer02.png
Could not find: null/doc/images/01archer03.png
Could not find: null/doc/images/01archer04.png
Could not find: null/doc/images/01archer.png
Could not find: null/doc/images/02axeman01.png
Could not find: null/doc/images/02axeman.png
Could not find: null/doc/images/03ballista01.png
Could not find: null/doc/images/04barbarian.png
Could not find: null/doc/images/05cataphract01.png
Could not find: null/doc/images/05cataphract02.png
Could not find: null/doc/images/06chariot01.png
Could not find: null/doc/images/07fort01.png
Could not find: null/doc/images/08horsearcher01.png
Could not find: null/doc/images/09horseman01.png
Could not find: null/doc/images/09horseman02.png
Could not find: null/doc/images/09horseman03.png
Could not find: null/doc/images/09horseman04.png
Could not find: null/doc/images/09horseman05.png
Could not find: null/doc/images/09horseman06.png
Could not find: null/doc/images/09horseman07.png
Could not find: null/doc/images/09horseman.png
Could not find: null/doc/images/10legionary01.png
Could not find: null/doc/images/10legionary02.png
Could not find: null/doc/images/11levis01.png
Could not find: null/doc/images/12onager01.png
Could not find: null/doc/images/13phalangite01.png
Could not find: null/doc/images/13phalangite02.png
Could not find: null/doc/images/13phalangite03.png
Could not find: null/doc/images/13phalangite04.png
Could not find: null/doc/images/14psilos01.png
Could not find: null/doc/images/14psilos02.png
Could not find: null/doc/images/14psilos03.png
Could not find: null/doc/images/15spearman01.png
Could not find: null/doc/images/15spearman02.png
Could not find: null/doc/images/15spearman03.png
Could not find: null/doc/images/15spearman04.png
Could not find: null/doc/images/15spearman.png
Could not find: null/doc/images/16swordman01.png
Could not find: null/doc/images/16swordman02.png
Could not find: null/doc/images/16swordman03.png
Could not find: null/doc/images/16swordman.png
Could not find: null/doc/images/17warelephant01.png
Could not find: null/doc/images/18quinquereme01.png
Could not find: null/doc/images/18quinquereme02.png
Could not find: null/doc/images/18quinquereme03.png
Could not find: null/doc/images/18quinquereme04.png
Could not find: null/doc/images/18quinquereme05.png
Could not find: null/doc/images/18quinquereme06.png
Could not find: null/doc/images/18quinquereme07.png
Could not find: null/doc/images/18quinquereme08.png
Could not find: null/doc/images/19trireme01.png
Could not find: null/doc/images/19trireme02.png
Could not find: null/doc/images/19trireme03.png
Could not find: null/doc/images/19trireme04.png
Could not find: null/doc/images/19trireme05.png
Could not find: null/doc/images/19trireme06.png
Could not find: null/doc/images/19trireme07.png
Could not find: null/doc/images/19trireme08.png
Could not find: null/doc/images/00city.png
```

That I assume is related to the fact that, apparently, I'm getting the game notes of the game I'm trying to join in the game I started with the interface I got while waiting to join:
![20180701_03](https://user-images.githubusercontent.com/19562310/42137379-9fcff43c-7d6b-11e8-8f6b-5ad389f825a1.png)


I get a double instance, and I can play one and spectate the other; so not actual issue, beside the error log and the gamenotes of the game I was joining, just wondering if you want to give users the chance to do something else with the same instance that they are waiting with for joining a game (and, in this case, then, there would be the issue of getting that game notes, as it happened this time).

**Comment by RoiEXLab** [comment]
@Cernelius I assume you have the "display error log" on everything setting active.
If you disable it, there shouldn't be any error message.

But your issue doesn't seem to have anything to do with this issue.
From my perspective it really seems to be a bad map configuration, but I'll have a look next week or so when I have more time. I'm happy that at least this issue is resolved for now.

**Comment by Cernelius** [comment]
Well, short story is that now, without the loading bar freezing the instance until successfully loaded, the program offers me the ability to bug myself if I'm fast enough to do something else in during the loading time.

**Comment by ssoloff** [comment]
@RoiEXLab I had some time today to look into the root cause of this issue, as requested via Gitter.  I think I have a plausible theory as to what's causing the `WaitDialog` to appear when a watcher leaves a bot-hosted game.  Hopefully it will give you some ideas as to how to ultimately solve the problem.

(Note that I investigated this issue based on `master` before #3502 was merged.  All snippets below reflect the source tree prior to that PR.)

#### TL;DR

Hiding the main frame window before hiding the `WaitDialog` spawned by `BackgroundTaskRunner` causes the `WaitDialog` to be shown when the main frame window is made visible after leaving a bot-hosted game.  The fix is to ensure the `WaitDialog` is hidden before hiding the main frame window.

#### Details

I was able to reliably reproduce the `WaitDialog` being shown when leaving a bot-hosted game.  The repro steps I followed are below for reference:

1. Run a local lobby.
1. Run a local bot connected to local lobby.
1. Start a new client and connect to local lobby as user `test1`.
    1. Join local bot.
    1. Select a map (I used Classic) and have user `test1` select all players.
    1. Wait for game to start.
    1. Play until start of combat move phase (see below).
1. Start a new client and connect to local lobby as user `test2`.
    1. Join local bot (since all players are selected, user `test2` is a guest/watcher).
    1. Wait for game to start.
1. From client for user `test1`, move one unit into combat.
    * :warning: Based on the theory I propose about `showWithParent` below, I'm not sure why this step is necessary.  If I skip this step (i.e. having user `test2` leave the game immediately after joining), the `WaitDialog` does not appear.  I didn't have time to investigate this further, but it bothers me that I had to do it to repro the issue.
1. From client for user `test2`, leave the game.

I attached a debugger to the `GameRunner` that is spawned when joining a remote game.  The problematic `WaitDialog` that is being displayed is originally created in `SetupPanelModel#showClient()` via a `BackgroundTaskRunner`:

```java
public void showClient(final Component ui) {
  final ClientModel model = new ClientModel(gameSelectorModel, this);
  Interruptibles.await(() -> GameRunner.newBackgroundTaskRunner().runInBackground("Loading game...", () -> {
    if (model.createClientMessenger(ui)) {
      SwingUtilities.invokeLater(() -> setGameTypePanel(new ClientSetupPanel(model)));
    } else {
      model.cancel();
    }
  }));
}
```

At some point later, **before `runInBackground()` returns**, the game is ready to start and `ClientModel#startGameInNewThread()` is called.  Within this method, the main frame window is hidden:

```java
    new Thread(() -> {
      SwingUtilities.invokeLater(() -> {
        JOptionPane.getFrameForComponent(ui).setVisible(false);
      });
      // ... snip ...
    }, "Client Game Launcher").start();
```

Stepping through the `setVisible()` call that hides the main frame window, I observed that the `Window#showWithParent` field of the `WaitDialog` is being set to `true` because the `WaitDialog` is still visible at this time.  This suggests that the `setVisible()` call is being made before the `SwingWorker#done()` method internal to the `BackgroundTaskRunner` is called.  I verified this by adding traces at these locations and confirmed the events are happening in the following order:

1. Show `WaitDialog`.
1. Hide main frame window.
1. Hide `WaitDialog`.

See [this SO question](https://stackoverflow.com/questions/25557774/java-not-disposing-dialog-when-frame-is-not-visible) for more information on how `showWithParent` affects child windows.

Thus, because `showWithParent` is `true`, when the main frame window is shown after leaving the game, the `WaitDialog` is also shown.  Because the actual task associated with the `WaitDialog` has already completed (i.e. the `SwingWorker`), there is no event that will trigger the re-shown `WaitDialog` to be closed a second time.  Since the dialog is modal, all input to the main frame window is blocked.

As a sanity check, I tried forcing the main frame window to be hidden _after_ the `WaitDialog` is hidden.  I simply moved the relevant code from `ClientModel#startGameInNewThread()` to `SetupPanelModel#showClient()`:

```diff
diff --git a/game-core/src/main/java/games/strategy/engine/framework/startup/mc/ClientModel.java b/game-core/src/main/java/games/strategy/engine/framework/startup/mc/ClientModel.java
index 4149c3c4b..7a5dbc981 100644
--- a/game-core/src/main/java/games/strategy/engine/framework/startup/mc/ClientModel.java
+++ b/game-core/src/main/java/games/strategy/engine/framework/startup/mc/ClientModel.java
@@ -338,7 +338,6 @@ public class ClientModel implements IMessengerErrorListener {
     final Messengers messengers = new Messengers(messenger, remoteMessenger, channelMessenger);
     game = new ClientGame(data, playerSet, players, messengers);
     new Thread(() -> {
-      SwingUtilities.invokeLater(() -> JOptionPane.getFrameForComponent(ui).setVisible(false));
       try {
         // game will be null if we loose the connection
         if (game != null) {
diff --git a/game-core/src/main/java/games/strategy/engine/framework/startup/mc/SetupPanelModel.java b/game-core/src/main/java/games/strategy/engine/framework/startup/mc/SetupPanelModel.java
index 559a428f0..c9a1d7f34 100644
--- a/game-core/src/main/java/games/strategy/engine/framework/startup/mc/SetupPanelModel.java
+++ b/game-core/src/main/java/games/strategy/engine/framework/startup/mc/SetupPanelModel.java
@@ -5,6 +5,7 @@ import java.awt.Dimension;
 import java.util.Observable;
 
 import javax.annotation.Nonnull;
+import javax.swing.JOptionPane;
 import javax.swing.JPanel;
 import javax.swing.SwingUtilities;
 
@@ -78,6 +79,7 @@ public class SetupPanelModel extends Observable {
         model.cancel();
       }
     }));
+    SwingUtilities.invokeLater(() -> JOptionPane.getFrameForComponent(ui).setVisible(false));
   }
 
   protected void setGameTypePanel(final ISetupPanel panel) {
```

With this change, I confirmed the event order from my original traces is now:

1. Show `WaitDialog`.
1. Hide `WaitDialog`.
1. Hide main frame window.

As expected, when the main frame window is shown after leaving the game, the `WaitDialog` no longer appears.

Note that I don't believe the simple change I made above is sufficient to fix the problem without introducing additional regressions.  There appears to be multiple code paths that go through `ClientModel#startGameInNewThread()` and rely on that method to hide the main frame window.  So  extracting that code to `SetupPanelModel#showClient()` will preclude the main frame window from being hidden on those other code paths because I don't believe they also go through the `showClient()` method.  (I only spent like five minutes looking at the other code paths, so my gut feeling about it not working could be totally wrong.)

Your theory about a race condition being present someplace is probably accurate, and would explain why not everyone experiences this problem.  This code is a total mess right now with all the `SwingUtilities#invokeLater()` calls and new `Thread`s being spawned haphazardly.  A proper fix might require redesigning the architecture of the code responsible for starting a game.

**Comment by prastle** [comment]
as a side note to help @ssoloff   and @RoiEXLab it occurred somewhere between releases10291 (works) and 10309 (doesn't work)

**Comment by DanVanAtta** [comment]
This still seems like an active topic, is this actually fixed? 
Re-opening for the moment, please close if this is actually resolved.

**Comment by DanVanAtta** [comment]
Also, perhaps this should be downgraded to a bug, looks like the repro is somewhat hard to be trigger (I could be wrong/might have missed something).

**Comment by RoiEXLab** [comment]
@DanVanAtta This has actually been "fixed" in the sense that this problem no longer occurs.
I removed the loading bar entirely to fix it for now, but ideally we'd come up with a better solution

**Comment by DanVanAtta** [comment]
@RoiEXLab is there an issue to track that we want to add the loading bars back? Should this be that ticket?

**Comment by RoiEXLab** [comment]
@DanVanAtta No there isn't. 
However you may want to read @ssoloff's post further up in this thread describing the cause of this issue.
It's going to be a bigger task to add them back. 
But it's probably a good idea to open an issue for this.

**Comment by DanVanAtta** [comment]
@RoiEXLab the reason I re-opened this was after reading @ssoloff's post above. It is some really good research and was well written. It does leave off open-ended, it seems like there is more to do. Let's please keep up the practice of team tracking and communication via github issues.
